axelrod/tests/unit/test_pickling.py

```diff
@@ -0,0 +1,349 @@
+import pickle
+import unittest
+
+import axelrod as axl
+import axelrod.strategy_transformers as st
+
+
+C, D = axl.Action.C, axl.Action.D
+
+
+# A set of classes to test pickling.
+
+
+PointerToWrappedStrategy = st.FlipTransformer()(
+    st.FlipTransformer()(axl.Cooperator)
+)
+
+
+class MyDefector(axl.Player):
+    def __init__(self):
+        super(MyDefector, self).__init__()
+
+    def strategy(self, opponent):
+        return D
+
+
+PointerToWrappedClassNotInStrategies = st.FlipTransformer()(
+    st.FlipTransformer()(MyDefector)
+)
+
+
+@st.FlipTransformer()
+class MyCooperator(axl.Player):
+    def strategy(self, opponent):
+        return C
+
+
+@st.FlipTransformer()
+@st.FlipTransformer()
+class DoubleFlip(axl.Cooperator):
+    pass
+
+
+@st.FlipTransformer()
+class SingleFlip(axl.Cooperator):
+    pass
+
+
+# All the transformers
+
+@st.ApologyTransformer([D], [C], name_prefix=None)
+class Apology(axl.Cooperator):
+    pass
+
+
+@st.DeadlockBreakingTransformer(name_prefix=None)
+class DeadlockBreaking(axl.Cooperator):
+    pass
+
+
+@st.DualTransformer(name_prefix=None)
+class Dual(axl.Cooperator):
+    pass
+
+
+@st.FlipTransformer(name_prefix=None)
+class Flip(axl.Cooperator):
+    pass
+
+
+@st.FinalTransformer((D, D), name_prefix=None)
+class Final(axl.Cooperator):
+    pass
+
+
+@st.ForgiverTransformer(0.2, name_prefix=None)
+class Forgiver(axl.Cooperator):
+    pass
+
+
+@st.GrudgeTransformer(3, name_prefix=None)
+class Grudge(axl.Cooperator):
+    pass
+
+
+@st.InitialTransformer((C, D), name_prefix=None)
+class Initial(axl.Cooperator):
+    pass
+
+
+@st.JossAnnTransformer((0.2, 0.2), name_prefix=None)
+class JossAnn(axl.Cooperator):
```

Comment:
There seem to be roughly 3 cases.

1) a decorated class declaration with no name prefix
    - based off of `Player`
    - based off of another strategy
2) a decorated class declaration with a name prefix
    - see `DoubleFlip` for an example
3) a dynamic wrapping of a class (with a name prefix)

i wanted to make sure that each transformer was tested on case-3 and either case-1 or case-2.  So these are module level class declarations for case-1, dynamically wrapped strategies for case-3, and some other module level class declarations for special cases and case-2.

and probably some overkill.  problems popped up in strange ways and i got paranoid.  pickling problems are weird and have little useful stack trace information when they throw an error.